signcrypt_seal.go

```diff
@@ -205,12 +204,29 @@ func shuffleSigncryptionReceivers(receiverBoxKeys []BoxPublicKey, receiverSymmet
 	return receivers
 }
 
+// signcryptRNG is an interface encapsulating all the randomness
+// (aside from ephemeral key generation) that happens during
+// signcryption. Tests can override it to make encryption
+// deterministic.
+type signcryptRNG interface {
+	createSymmetricKey() (*SymmetricKey, error)
+	shuffleReceivers(receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey) []receiverKeysMaker
+}
+
 // This generates the payload key, and encrypts it for all the different
 // recipients of the two different types. Symmetric key recipients and DH key
 // recipients use different types of identifiers, but they are the same length,
 // and should both be indistinguishable from random noise.
-func (sss *signcryptSealStream) init(receivers []receiverKeysMaker) error {
-	ephemeralKey, err := sss.keyring.CreateEphemeralKey()
+func (sss *signcryptSealStream) init(
+	receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey,
+	ephemeralKeyCreator EphemeralKeyCreator, rng signcryptRNG) error {
+	if len(receiverBoxKeys)+len(receiverSymmetricKeys) == 0 {
```

Comment:
I forget...is there a reason why we're not checking receiver uniqueness here, like we do for encryption? (and also for too many receivers...)

Reply:
I don't think so. Maybe we should just copy the `checkReceivers` function wholesale?

Reply:
Done. Also adjusted max receiver count to account for other fields in the array.